<TITLE>Gene Model Loads</TITLE>

<H1>Ensembl</H1>
<UL>

<H2>Data</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/input/ensembl_genemodels.txt">Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/input/ensembl_assoc.txt">Association Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/ensembl_genemodels_load.txt">Load-Ready Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/ensembl_assoc_load.txt">Load-Ready Association Input File</A>
</UL>

<H2>Logs</H4>
<UL>

<H4>Gene Model Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/ensembl_genemodelQC.log">Sanity/QC Log</A>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/ensembl_genemodelload.log">Gene Model Load Log</A>
</UL>

<H4>Assembly Sequence Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ensemblseqload/logs/assemblyseqload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/ensemblseqload/logs/assemblyseqload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/ensemblseqload/logs/assemblyseqload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/ensemblseqload/logs/assemblyseqload.val.log">Data Validation Log</A>
</UL>

<H4>Association Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/logs/assocload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/logs/assocload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/logs/assocload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/logs/assocload.val.log">Data Validation Log</A>
</UL>
</UL>

<H2>Reports</H4>
<UL>

<H4>Pre-Load Sanity/QC Reports</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_genemodels_sanity.rpt">Gene Model Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_assoc_sanity.rpt">Association Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_invalid_marker.rpt">Invalid Marker Report</A> - MGI IDs in the association file that are not accession IDs for a valid marker
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_sec_marker.rpt">Secondary Marker Report</A> - MGI IDs in the association file that are secondary marker IDs
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_missing_gmid.rpt">Missing Gene Model ID Report</A> - Gene Model IDs in the association file that are not in the gene model file
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ensembl_chr_discrep.rpt">Chromosome Discrepancy Report</A> - Gene models that have a different chromosome than the marker that they are being associated with
</UL>

<H4>Association Load Reports</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/reports/TargetDiscrepancy.rpt">MGI Marker Discrepancies</A>
<LI><A HREF="/data/loads/assembly/ensembl_assocload/reports/AssocDiscrepancy.rpt">Ensembl Gene Model Discrepancies</A>
</UL>
</UL>
<H2>VEGA/Ensembl Transcript/Protein Sequence and Marker Association Loads</H4>
<UL>
<LI><A HREF="http://firien.informatics.jax.org/usrlocalmgi/live/dataload/vega_ensemblseqload/index.html">Index File</A>
</UL>
</UL>

<H1>NCBI</H1>
<UL>

<H2>Data</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/input/ncbi_genemodels.txt">Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/input/ncbi_assoc.txt">Association Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/ncbi_genemodels_load.txt">Load-Ready Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/ncbi_assoc_load.txt">Load-Ready Association Input File</A>
</UL>

<H2>Logs</H4>
<UL>

<H4>Gene Model Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/ncbi_genemodelQC.log">Sanity/QC Log</A>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/ncbi_genemodelload.log">Gene Model Load Log</A>
</UL>

<H4>Assembly Sequence Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ncbiseqload/logs/assemblyseqload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/ncbiseqload/logs/assemblyseqload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/ncbiseqload/logs/assemblyseqload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/ncbiseqload/logs/assemblyseqload.val.log">Data Validation Log</A>
</UL>

<H4>Association Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/logs/assocload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/logs/assocload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/logs/assocload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/logs/assocload.val.log">Data Validation Log</A>
</UL>
</UL>

<H2>Reports</H4>
<UL>

<H4>Pre-Load Sanity/QC Reports</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_genemodels_sanity.rpt">Gene Model Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_assoc_sanity.rpt">Association Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_invalid_marker.rpt">Invalid Marker Report</A> - MGI IDs in the association file that are not accession IDs for a valid marker
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_sec_marker.rpt">Secondary Marker Report</A> - MGI IDs in the association file that are secondary marker IDs
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_missing_gmid.rpt">Missing Gene Model ID Report</A> - Gene Model IDs in the association file that are not in the gene model file
<LI><A HREF="/data/loads/mgi/genemodelload/reports/ncbi_chr_discrep.rpt">Chromosome Discrepancy Report</A> - Gene models that have a different chromosome than the marker that they are being associated with
</UL>

<H4>Association Load Reports</H4>
<UL>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/reports/TargetDiscrepancy.rpt">MGI Marker Discrepancies</A>
<LI><A HREF="/data/loads/assembly/ncbi_assocload/reports/AssocDiscrepancy.rpt">NCBI Gene Model Discrepancies</A>
</UL>
</UL>
</UL>

<H1>VEGA</H1>
<UL>

<H2>Data</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/input/vega_genemodels.txt">Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/input/vega_assoc.txt">Association Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/vega_genemodels_load.txt">Load-Ready Gene Model Input File</A>
<LI><A HREF="/data/loads/mgi/genemodelload/output/vega_assoc_load.txt">Load-Ready Association Input File</A>
</UL>

<H2>Logs</H4>
<UL>

<H4>Gene Model Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/vega_genemodelQC.log">Sanity/QC Log</A>
<LI><A HREF="/data/loads/mgi/genemodelload/logs/vega_genemodelload.log">Gene Model Load Log</A>
</UL>

<H4>Assembly Sequence Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/vegaseqload/logs/assemblyseqload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/vegaseqload/logs/assemblyseqload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/vegaseqload/logs/assemblyseqload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/vegaseqload/logs/assemblyseqload.val.log">Data Validation Log</A>
</UL>

<H4>Association Load Logs</H4>
<UL>
<LI><A HREF="/data/loads/assembly/vega_assocload/logs/assocload.proc.log">Process Summary Log</A>
<LI><A HREF="/data/loads/assembly/vega_assocload/logs/assocload.diag.log">Diagnostic Log</A>
<LI><A HREF="/data/loads/assembly/vega_assocload/logs/assocload.cur.log">Curator Summary Log</A>
<LI><A HREF="/data/loads/assembly/vega_assocload/logs/assocload.val.log">Data Validation Log</A>
</UL>
</UL>

<H2>Reports</H4>
<UL>

<H4>Pre-Load Sanity/QC Reports</H4>
<UL>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_genemodels_sanity.rpt">Gene Model Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_assoc_sanity.rpt">Association Input File Sanity Report</A>
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_invalid_marker.rpt">Invalid Marker Report</A> - MGI IDs in the association file that are not accession IDs for a valid marker
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_sec_marker.rpt">Secondary Marker Report</A> - MGI IDs in the association file that are secondary marker IDs
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_missing_gmid.rpt">Missing Gene Model ID Report</A> - Gene Model IDs in the association file that are not in the gene model file
<LI><A HREF="/data/loads/mgi/genemodelload/reports/vega_chr_discrep.rpt">Chromosome Discrepancy Report</A> - Gene models that have a different chromosome than the marker that they are being associated with
</UL>

<H4>Association Load Reports</H4>
<UL>
<LI><A HREF="/data/loads/assembly/vega_assocload/reports/TargetDiscrepancy.rpt">MGI Marker Discrepancies</A>
<LI><A HREF="/data/loads/assembly/vega_assocload/reports/AssocDiscrepancy.rpt">VEGA Gene Model Discrepancies</A>
</UL>
</UL>
<H2>VEGA/Ensembl Transcript/Protein Sequence and Marker Association Loads</H4>
<UL>
<LI><A HREF="http://firien.informatics.jax.org/usrlocalmgi/live/dataload/vega_en
semblseqload/index.html">Index File</A>
</UL>
</UL>
